build.zig

```diff
@@ -67,20 +67,79 @@ pub fn build(b: *Builder) !void {
-        .aarch64 =>
-        // armstub not yet working, therefore: b 0x80000 (which is 0x14020000) followed by 0 filler until 0x80000 followed by kernel that starts at 0x80000
-        b.addSystemCommand(&[_][]const u8{
-            "/bin/bash",
-            "-c",
-            try std.fmt.allocPrint(b.allocator, "aarch64-linux-gnu-objcopy {} -O binary {} && mkdir --parents {} && cp --archive src/kernel/arch/aarch64/rpi-sdcard/config.txt {} && echo -ne \"\\x00\\x00\\x02\\x14\" > {} && dd status=none bs=1 count=524284 if=/dev/zero >> {} && cat {}/kernel8.img >> {} && wget --directory-prefix={} --quiet --timestamp {}/bootcode.bin {}/fixup.dat {}/start.elf && zip --junk-paths --quiet --recurse-paths {}.zip {}", .{ exec.getOutputPath(), image_file_name, zip_folder, zip_folder, kernel, kernel, b.cache_root, kernel, zip_folder, firmware_url, firmware_url, firmware_url, zip_folder, zip_folder }),
-        }),
+        .aarch64 => zipSequence: {
+            const zip_folder = try fs.path.join(b.allocator, &[_][]const u8{ b.cache_root, "rpi-sdcard" });
+            const elf = try fs.path.join(b.allocator, &[_][]const u8{ exec.output_dir.?, "pluto.elf" });
+            const firmware_version_tag = "1.20200601+arm64";
+            const firmware_url = "https://github.com/raspberrypi/firmware/raw/" ++ firmware_version_tag ++ "/boot/";
+            const kernel = try fs.path.join(b.allocator, &[_][]const u8{ zip_folder, "kernel8.img" });
+
+            const mkdir = b.addSystemCommand(&[_][]const u8{
+                "mkdir",
+                "--parents",
+                zip_folder,
+            });
+            mkdir.step.dependOn(&exec.step);
+
+            const objcopy = b.addSystemCommand(&[_][]const u8{
+                "aarch64-linux-gnu-objcopy",
+                elf,
+                "-O",
+                "binary",
+                try fs.path.join(b.allocator, &[_][]const u8{ b.cache_root, "kernel8.img" }),
+            });
+            objcopy.step.dependOn(&mkdir.step);
+
+            const cp_config = b.addSystemCommand(&[_][]const u8{
+                "cp",
+                "--archive",
+                "src/kernel/arch/aarch64/rpi-sdcard/config.txt",
+                zip_folder,
+            });
+            cp_config.step.dependOn(&objcopy.step);
+
+            const emulate_armstub = armstubSequence: {
+                // armstub not yet working, therefore: b 0x80000 (which is 0x14020000)
+                const create_b_0x80000 = bash(b, "echo -ne \"\\x00\\x00\\x02\\x14\" > {}", .{kernel});
+                create_b_0x80000.step.dependOn(&cp_config.step);
+
+                // followed by 0 filler until 0x80000
+                const dd = bash(b, "dd status=none bs=1 count=524284 if=/dev/zero >> {}", .{kernel});
+                dd.step.dependOn(&create_b_0x80000.step);
+
+                // followed finally by kernel that starts at 0x80000
+                const cat = bash(b, "cat {}/kernel8.img >> {}", .{ b.cache_root, kernel });
```

Comment:
Why is this necessary? Shouldn't it just be a case of packaging the kernel image, config etc. into a .img file?

Reply:
This is part of putting the image at 0x80000 with a jump instruction at 0. The bash has been replaced by zig.

https://github.com/markfirmware/pluto/blob/0fb94376e5622874046f0d874f15966c373f59fa/build.zig#L214-L242

Reply:
The entire disk is now packaged into rpi-sdcard.img.

Reply:
@SamTebbs33